File: compiler/nodes/expressions/constructor.go
package expressions

import (
	"reflect"
	"slices"

	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type ConstructorExpressionNode struct {
	ConstructorName string
	Facts           *database.Facts

	matchingConstructorDefinition visit.Definition
	matchingSubstitutions         map[database.Node]typecheck.Type
}

func (node *ConstructorExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseConstructorExpression(parser *syntax.Parser) (*ConstructorExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseConstructorName(parser)
	if err != nil {
		return nil, err
	}

	return &ConstructorExpressionNode{
		ConstructorName: name,
		Facts:           database.NewFacts(span()),
	}, nil
}

func (node *ConstructorExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	constructorDefinition, ok := visit.ResolveOf(visitor, node.ConstructorName, node, []reflect.Type{
		reflect.TypeFor[*visit.TraitDefinition](),
		reflect.TypeFor[*visit.VariantConstructorDefinition](),
		reflect.TypeFor[*visit.MarkerConstructorDefinition](),
	})

	if !ok {
		return
	}

	substitutions := map[database.Node]typecheck.Type{}

	visitor.Constraint(typecheck.NewInstantiateConstraint(typecheck.Instantiation{
		Source:        node,
		Definition:    constructorDefinition.GetNode(),
		Substitutions: &substitutions,
		Replacements:  map[database.Node]database.Node{constructorDefinition.GetNode(): node},
	}, visit.GetDefinitionConstraints))

	if traitDefinition, ok := constructorDefinition.(*visit.TraitDefinition); ok {
		visitor.Constraint(typecheck.NewBoundConstraint(node, typecheck.UnresolvedBound{
			Source:        node,
			Trait:         traitDefinition.Node,
			Substitutions: &substitutions,
			TraitName:     traitDefinition.Name,
		}, visit.GetDefinitionConstraints))

		node.matchingSubstitutions = substitutions
	}

	node.matchingConstructorDefinition = constructorDefinition
}

func (node *ConstructorExpressionNode) Codegen(c *codegen.Codegen) error {
	if node.matchingConstructorDefinition == nil {
		return c.Error(node)
	}

	span := database.GetSpanFact(node)

	switch definition := node.matchingConstructorDefinition.(type) {
	case *visit.MarkerConstructorDefinition:
		c.WriteString(span, "null")
	case *visit.TraitDefinition:
		if node.matchingSubstitutions == nil {
			return c.Error(node)
		}

		parameters := make([]database.Node, 0, len(node.matchingSubstitutions))
		for parameter := range node.matchingSubstitutions {
			parameters = append(parameters, parameter)
		}
		slices.SortStableFunc(parameters, func(left database.Node, right database.Node) int {
			return database.CompareSpans(database.GetSpanFact(left), database.GetSpanFact(right))
		})

		c.WriteString(span, "await __wipple_trait(")
		c.WriteNode(span, definition.Node)
		c.WriteString(span, ", __wipple_types, {")

		for _, parameter := range parameters {
			ty := node.matchingSubstitutions[parameter]
			c.WriteNode(span, parameter)
			c.WriteString(span, ": ")
			if err := c.WriteType(span, ty); err != nil {
				return err
			}
			c.WriteString(span, ", ")
		}

		c.WriteString(span, "})")
	case *visit.VariantConstructorDefinition:
		c.Write(definition.Node)
	default:
		return c.Error(node)
	}

	return nil
}

File: compiler/src/nodes/expressions/constructor.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Fact, HiddenNode, Node, NodeRef, Render},
    nodes::{codegen_instance, visit_expression},
    syntax::{ParseError, Parser, parse_constructor_name},
    typecheck::{
        Bound, BoundConstraint, Bounds, Instantation, InstantiateConstraint, Replacements,
        Substitutions, TypeConstraint,
    },
    visit::{Definition, VariantConstructorDefinition, Visit, Visitor},
};

#[derive(Debug, Clone)]
pub enum ResolvedConstructor {
    Marker,
    Trait(bool),
    Variant(VariantConstructorDefinition, NodeRef),
}

impl Fact for ResolvedConstructor {}

impl Render for ResolvedConstructor {}

#[derive(Debug)]
pub struct ConstructorExpressionNode {
    pub constructor: String,
}

impl Node for ConstructorExpressionNode {}

pub fn parse_constructor_expression(
    parser: &mut Parser<'_>,
) -> Result<ConstructorExpressionNode, ParseError> {
    let constructor = parse_constructor_name(parser)?;

    Ok(ConstructorExpressionNode { constructor })
}

impl Visit for ConstructorExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        let Some(definition) =
            visitor.resolve(&self.constructor, node, |definition| match definition {
                Definition::Trait(definition) => Some(Definition::Trait(definition.clone())),
                Definition::VariantConstructor(definition) => {
                    Some(Definition::VariantConstructor(definition.clone()))
                }
                Definition::MarkerConstructor(definition) => {
                    Some(Definition::MarkerConstructor(definition.clone()))
                }
                _ => None,
            })
        else {
            return;
        };

        visitor.graph.instantiate(node, &definition.node(), node);

        let substitutions = Substitutions::new();

        let definition_node = definition.node();
        visitor.constraint(InstantiateConstraint::new(Instantation {
            source_node: node.clone(),
            definition: definition_node.clone(),
            substitutions: substitutions.clone(),
            replacements: Replacements::from_iter([(definition_node, node.clone())]),
            from_expression: true,
        }));

        match definition {
            Definition::Trait(trait_definition) => {
                visitor.constraint(BoundConstraint::new(
                    node.clone(),
                    Bound {
                        source_node: node.clone(),
                        bound_node: node.clone(),
                        trait_node: trait_definition.node,
                        target_node: None,
                        substitutions,
                        optional: false,
                    },
                ));

                let generic = visitor.try_current_definition().is_some();

                visitor.insert(node, ResolvedConstructor::Trait(generic));
            }
            Definition::VariantConstructor(definition) => {
                let result = if definition.elements.is_empty() {
                    node.clone()
                } else {
                    let span = visitor.span(node);

                    let element_temporaries = definition
                        .elements
                        .iter()
                        .map(|node| {
                            let span = visitor.span(node);
                            visitor.node(span, HiddenNode::default())
                        })
                        .collect::<Vec<_>>();

                    let result_temporary = visitor.node(span, HiddenNode::default());

                    visitor.constraint(TypeConstraint::new(
                        node.clone(),
                        visitor.function_type(element_temporaries, result_temporary.clone()),
                    ));

                    result_temporary
                };

                visitor.insert(node, ResolvedConstructor::Variant(definition, result));
            }
            Definition::MarkerConstructor(_) => {
                visitor.insert(node, ResolvedConstructor::Marker);
            }
            _ => {}
        }
    }
}

impl Codegen for ConstructorExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let constructor = ctx
            .get::<ResolvedConstructor>(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        match constructor {
            ResolvedConstructor::Marker => {
                ctx.instruction(ir::Instruction::Value {
                    node: node.clone(),
                    value: ir::Value::Marker,
                });
            }
            ResolvedConstructor::Trait(generic) => {
                let bounds = ctx.get::<Bounds>(node).unwrap_or_default();

                let instance = bounds
                    .0
                    .get(node)
                    .and_then(|bound| bound.instance.as_ref())
                    .ok_or_else(|| anyhow::format_err!("unresolved"))?;

                let instance = codegen_instance(ctx, instance, generic)?;

                match instance {
                    ir::Instance::Bound(bound) => {
                        ctx.instruction(ir::Instruction::Value {
                            node: node.clone(),
                            value: ir::Value::Bound(bound),
                        });
                    }
                    ir::Instance::Definition(key) => {
                        ctx.instruction(ir::Instruction::Value {
                            node: node.clone(),
                            value: ir::Value::Constant(key),
                        });
                    }
                }
            }
            ResolvedConstructor::Variant(definition, result) => {
                if definition.elements.is_empty() {
                    ctx.instruction(ir::Instruction::Value {
                        node: node.clone(),
                        value: ir::Value::Variant {
                            index: definition.index,
                            elements: Vec::new(),
                        },
                    });
                } else {
                    ctx.instruction(ir::Instruction::Value {
                        node: node.clone(),
                        value: ir::Value::Function {
                            inputs: definition.elements.clone(),
                            captures: Vec::new(),
                            instructions: vec![
                                ir::Instruction::Value {
                                    node: result.clone(),
                                    value: ir::Value::Variant {
                                        index: definition.index,
                                        elements: definition.elements,
                                    },
                                },
                                ir::Instruction::Return { value: result },
                            ],
                        },
                    });
                }
            }
        }

        Ok(())
    }
}
